### PR TITLE
MM-21328: Fix KVCompareAndSet when new==old

### PR DESCRIPTION
**Pull request by lieut-data**
#### Summary
`KVCompareAndSet(key, sameValue, sameValue)` can fail spuriously on MySQL if the underlying `UPDATE` requires no actual changes. As per the [MySQL documentation](https://dev.mysql.com/doc/refman/8.0/en/mysql-affected-rows.html), we can't rely on rows affected in this case:

> For UPDATE statements, the affected-rows value by default is the number of rows actually changed. If you specify the CLIENT_FOUND_ROWS flag to mysql_real_connect() when connecting to mysqld, the affected-rows value is the number of rows “found”; that is, matched by the WHERE clause.

It's not tenable to change `CLIENT_FOUND_ROWS` for all connections, so handle this case in the code instead by running a `SELECT` after the fact. Note that `KVCompareAndSet` has no guarantee of atomicity in this case, but neither would `CompareAndSwap` on which this is method was inspired.

Finally, note that no changes are required for Postgres, which has sane semantics as the default.

#### Ticket Link
Fixes: https://mattermost.atlassian.net/browse/MM-21328

### CODE REVIEW

File: store/sqlstore/plugin_store.go [hunk 1]
```diff
@@ -108,6 +109,33 @@ func (ps SqlPluginStore) CompareAndSet(kv *model.PluginKeyValue, oldValue []byte
...
+					return false, model.NewAppError("SqlPluginStore.CompareAndSet", "store.sql_plugin_store.compare_and_set.mysql_select.app_error", nil, fmt.Sprintf("plugin_id=%v, key=%v, err=%v", kv.PluginId, kv.Key, err.Error()), http.StatusInternalServerError)
+				}
+
+				if count == 0 {
+					return false, nil
+				} else if count == 1 {
+					return true, nil
+				} else {
+					return false, model.NewAppError("SqlPluginStore.CompareAndSet", "store.sql_plugin_store.compare_and_set.too_many_rows.app_error", nil, fmt.Sprintf("plugin_id=%v, key=%v, count=%d", kv.PluginId, kv.Key, count), http.StatusInternalServerError)
+				}
```
Comment (hanzei) [inline store/sqlstore/plugin_store.go]: Nit: Use `switch` instead

File: store/sqlstore/plugin_store.go [hunk 2]
```diff
@@ -108,6 +109,33 @@ func (ps SqlPluginStore) CompareAndSet(kv *model.PluginKeyValue, oldValue []byte
...
+				if err != nil {
+					return false, model.NewAppError("SqlPluginStore.CompareAndSet", "store.sql_plugin_store.compare_and_set.mysql_select.app_error", nil, fmt.Sprintf("plugin_id=%v, key=%v, err=%v", kv.PluginId, kv.Key, err.Error()), http.StatusInternalServerError)
+				}
+
+				if count == 0 {
+					return false, nil
+				} else if count == 1 {
+					return true, nil
+				} else {
+					return false, model.NewAppError("SqlPluginStore.CompareAndSet", "store.sql_plugin_store.compare_and_set.too_many_rows.app_error", nil, fmt.Sprintf("plugin_id=%v, key=%v, count=%d", kv.PluginId, kv.Key, count), http.StatusInternalServerError)
```
Comment (hanzei) [inline store/sqlstore/plugin_store.go]: Just to double check: This should _never_ happen, right?

Comment (lieut-data) [reply store/sqlstore/plugin_store.go]: Correct: the primary keys, if in place, guarantee the database would never allow this situation to exist.

File: store/sqlstore/plugin_store.go (hunk 1 above)
Comment (lieut-data) [reply store/sqlstore/plugin_store.go]: 1/5, but opting to leave as-is unless objection is stronger.

Comment (hanzei) [reply store/sqlstore/plugin_store.go]: I'm also 1/5, so it can stay this way

**Comment by prapti** [comment]
More tests will be done after merging.

**Pull request merged**